File: DotNet/Messaging/Storage/PreferenceOrderProvider.cs
using System.Collections.Generic;
using System.Linq;

namespace Bollywell.Hydra.Messaging.Storage
{
    /// <summary>
    /// Use the first server in the preference list that is responding. Switch if a earlier one in the list becomes available.
    /// </summary>
    public class PreferenceOrderProvider : PollingProviderBase
    {
        private readonly List<string> _servers;
        private readonly Dictionary<string, int> _serverIndices = new Dictionary<string, int>();
        private int _preferredIndex;

        /// <summary>
        /// Initialise messaging. Must be called before any attempt to send or listen.
        /// </summary>
        /// <param name="hydraServers">Hydra servers to communicate with, in decreasing order of preference</param>
        /// <param name="database">Name of the messaging database. Defaults to "hydra"</param>
        /// <param name="port">Port number of the messaging database. Defaults to 5984</param>
        public PreferenceOrderProvider(IEnumerable<string> hydraServers, string database = DefaultDatabase, int port = DefaultPort)
            : this(hydraServers.Select(s => new CouchDbStore(string.Format("{0}:{1}:{2}", s, port, database), s, database, port))) { }

        /// <summary>
        /// Initialise messaging. Must be called before any attempt to send or listen.
        /// </summary>
        /// <param name="stores">Hydra stores to communicate with</param>
        public PreferenceOrderProvider(IEnumerable<IStore> stores) : base(stores)
        {
            _servers = stores.Select(s => s.Name).ToList();
            Enumerable.Range(0, _servers.Count).ToList().ForEach(index => _serverIndices[_servers[index]] = index);
            _preferredIndex = _servers.Count;
        }

        protected override void OnDistanceInfo(ServerDistanceInfo sdi)
        {
            var sdiIndex = _serverIndices[sdi.Name];
            if (sdi.IsReachable && sdiIndex < _preferredIndex
                && _servers.Take(sdiIndex - 1).All(s => Distances.ServerInfo.ContainsKey(s) && !Distances.ServerInfo[s].IsReachable)) {
                // A server is preferred if it is earlier in _servers than the current server, it is reachable, and all servers before it in _servers are unreachable.
                _preferredIndex = sdiIndex;
                HydraServer = sdi.Name;
            } else if (!sdi.IsReachable && sdiIndex == _preferredIndex) {
                // The current server is no longer responding - replace with the first reachable one before which all others are unreachable, or else null
                var candidate = _servers.SkipWhile(s => Distances.ServerInfo.ContainsKey(s) && !Distances.ServerInfo[s].IsReachable).FirstOrDefault();
                HydraServer = candidate != null && Distances.ServerInfo.ContainsKey(candidate) ? candidate : null;
                _preferredIndex = HydraServer == null ? _servers.Count : _serverIndices[HydraServer];
            }
        }

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/storage/PreferenceOrderProvider.java
package uk.co.shastra.hydra.messaging.storage;

import java.util.ArrayList;
import java.util.HashMap;

import rx.Observable;
import rx.util.functions.Action1;

public class PreferenceOrderProvider extends PollingProviderBase {

    private ArrayList<String> servers = new ArrayList<String>();
    private HashMap<String, Integer> serverIndices = new HashMap<String, Integer>();
    private int preferredIndex;
    
    public PreferenceOrderProvider(Iterable<String> hydraServers, String database) { this(hydraServers, database, null); }
    public PreferenceOrderProvider(Iterable<String> hydraServers, String database, Integer port) { 
    	super(hydraServers, database, port);
    	init(hydraServers);
    }

	public PreferenceOrderProvider(Iterable<Store> stores) { 
    	super(stores);
    	ArrayList<String> hydraServers = new ArrayList<String>();
    	for (Store store : stores) {
			hydraServers.add(store.getName());
		}
    	init(hydraServers);
    }

    private void init(Iterable<String> hydraServers) {
    	int count = 0;
    	for (String server : hydraServers) {
    		servers.add(server);
			serverIndices.put(server, count);
		}
    	preferredIndex = count;
	}
    
	@Override
	protected void doSubscription(Observable<ServerDistanceInfo> observable) {
		observable.subscribe(new Action1<ServerDistanceInfo>() {
			@Override
			public void call(ServerDistanceInfo sdi) {
	            int sdiIndex = serverIndices.get(sdi.name);
	            if (sdi.isReachable && sdiIndex < preferredIndex && EarlierServersUnreachable(sdiIndex)) {
	                // A server is preferred if it is earlier in servers than the current server, it is reachable, and all servers before it in servers are unreachable.
	                preferredIndex = sdiIndex;
	                setHydraServer(sdi.name);
	            } else if (!sdi.isReachable && sdiIndex == preferredIndex) {
	                // The current server is no longer responding - replace with the first reachable one before which all others are unreachable, or else null
	            	String server = null;
	                int ii = 0;
	                while (ii < servers.size()) {
	                	String candidate = servers.get(ii);
	                	if (distances.getServerInfo().containsKey(candidate) && !distances.getServerInfo().get(candidate).isReachable) {
	                		ii++;
	                	} else {
	                		server = candidate;
	                		break;
	                	}
					}
	                setHydraServer(server != null && distances.getServerInfo().containsKey(server) ? server : null);
	                preferredIndex = getHydraServer() == null ? servers.size() : serverIndices.get(getHydraServer());
	            } else if (!initialised && !sdi.isReachable && distances.getServerInfo().size() == storeDict.size()) {
	                // All servers are unreachable so initialisation is done
	                finishedInitialisation();
	            }
			}

			/**
			 * Check whether all servers before sdiIndex are in distance.getServerInfo() and are unreachable
			 * 
			 * @param sdiIndex
			 * @return Whether the condition is true
			 */
			private boolean EarlierServersUnreachable(int sdiIndex) {
				for (int ii = 0; ii < sdiIndex; ii++) {
					String server = servers.get(ii);
					if (!distances.getServerInfo().containsKey(server) || distances.getServerInfo().get(server).isReachable)
						return false;
				}
				return true;
			}
		});
	}

}
